cloudinit/sources/DataSourceEc2.py

```diff
@@ -260,6 +260,8 @@ def wait_for_metadata_service(self):
         if metadata_address:
             self.metadata_address = metadata_address
             LOG.debug("Using metadata source: '%s'", self.metadata_address)
+        elif self.cloud_name == CloudNames.AWS:
```

Comment:
If we end up not finding a metadata_address in _wait_for_metadata_service then cloud-init will return an empty crawled_metadata dict in line 422 or def crawl_metadata. This would result in cloud-init saying that DataSourceEc2 is an invalid datasource for this vm and it would move on to other datasources. We want to always ensure we fallback to IMDSv1 even on ec2 lookalikes, if anything, this should b3 critical not a warning I'd say as I expect we'll not detect any datasource on this type of machine if we've gotten this far and still don't have a metadata_address

Reply:
Are there ec2 lookalikes for which ```CloudNames.AWS``` would be **True**?

To answer the 3 comments at once, I'm fine with checking specifically for the 403 error. Handling a timeout on the request for a new token is not a bad idea for falling back to IMDSv1 at that point is not desirable. On EC2, if the token request times out, we should retry ; going back to my question above about CloudNames.AWS and lookalikes.

I'll update the patch

Reply:
Thanks @fred-lefebvre for that, thanks for limiting the check to 403 errors.  I look forward to the changeset you mentioned.

To answer the question about Ec2-lookalikes, I'm not aware of any cloud platform the 'fakes'  being strictly CloudNames.AWS. The only way do do that would be for the platform to surface a DMI system-serial-number, chassis-asset-tag and system-serial-number to meet the following identitfication criteria: 

```
def identify_aws(data):
    # data is a dictionary returned by _collect_platform_data.
    if (data['uuid'].startswith('ec2') and
            (data['uuid_source'] == 'hypervisor' or
             data['uuid'] == data['serial'])):
        return CloudNames.AWS

    return None
```

If a platforms does fake DMI data to look like CloudNames.AWS, then they are beholden to feature sets and changes that true AWS surfaces, so they should be able to cope with IMDSv2 requirements. If they have special requirements otherwise that prevent them from running, then we'll need them to identify properly as a different ec2-lookalike CloudName.

So I think you don't have to worry about non-AWS ec2 lookalikes specifically here.
I was more concerned that AWS vms, on ec2-proper would  drop through during a timeout or unexpected error in  self._maybe_fetch_api_token() and skip straight to not detecting DataSourceEc2 at all if there were an issue with the IMDSv2 service being accessible during the vm boot and cloud-init datasource discovery.

Reply:
@blackboxsw : Thanks for the review.  Just to be clear, does  *"I look forward to the changeset you mentioned"* refer to my earlier comment about addressing the recovery when no data source can be discovered on an already configured instance?  I intend to take this on next but I would argue that this current change set can stand on its own and should be committed by itself to keep commits small and easier to review.

With the change here we improve IMDSv2 support by providing better logging  and a faster abort when the http endpoint is disabled.  Restricting the change to ```CloudNames.AWS``` makes it fairly safe.  Admittedly, the code here suffers from the same failings as the current cloud-init code base : If the IMDS endpoint is disabled, we are left without the ```/var/lib/cloud/instance``` symlink. But at least we get there faster. I would like to address the consistency issue in a separate PR as it will impact code outside of the EC2 datasource and will be easier to review on its own.  

**In short:** What do you want done before merging this PR?